File: leetcode/Tree/leetcode-114/114.go
package tree

import "container/list"

// TreeNode Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func flatten(root *TreeNode) {
	if root == nil {
		return
	}

	stack := list.New()
	node, pre := root, &TreeNode{0, nil, nil}
	for node != nil {
		stack.PushBack(node)
		pre.Left = node
		pre = pre.Left
		node = node.Left
	}

	for stack.Len() != 0 {
		node = stack.Back().Value.(*TreeNode)
		stack.Remove(stack.Back())
		if node.Right != nil {
			tmp := node.Right
			node.Right = nil
			node = tmp
			for node != nil {
				stack.PushBack(node)
				pre.Left = node
				pre = pre.Left
				node = node.Left
			}
		}
	}

	node = root
	for node != nil {
		node.Right = node.Left
		node.Left = nil
		node = node.Right
	}
}

File: leetcode/Tree/leetcode-114/114.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        if (root == null) {
            return;
        }

        Stack<TreeNode> stack = new Stack<>();
        TreeNode pre = new TreeNode(0);
        TreeNode node = root;
        while (node != null) {
            stack.push(node);
            pre.left = node;
            pre = pre.left;
            node = node.left;
        }

        while (!stack.isEmpty()) {
            node = stack.pop();
            if (node.right != null) {
                TreeNode tmp = node.right;
                node.right = null;
                node = tmp;
                while (node != null) {
                    stack.push(node);
                    pre.left = node;
                    pre = pre.left;
                    node = node.left;
                }
            }          
        }

        node = root;
        while (node != null) {
            node.right = node.left;
            node.left = null;
            node = node.right;
        }
    }
}